### PR TITLE
Resource formulas: compute exec memory/CPU from input file metrics

### PR DESCRIPTION
**Pull request by PaulNewling**
Code, tests, and review fixes are complete and verified end-to-end against a live backend. Merge is gated only on the `@platforma-sdk/workflow-tengo` release and three dependent block PRs (see Status).

## Summary

Block workflows can size an exec's RAM and CPU from its input data instead of hardcoding `.mem("64GiB").cpu(16)`, which either over-allocates or OOMs. A new `exec.formula` DSL builds an expression AST; `exec.builder()` gains `.memFormula()` / `.cpuFormula()` and a `{ tag }` on `.addFile()`. The formula evaluates at exec time from input-file metrics.

```go
f := exec.formula
exec.builder()
    .addFile("reads_R1.fastq.gz", r1, { tag: "reads" })
    .addFile("reads_R2.fastq.gz", r2, { tag: "reads" })
    .memFormula(f.clamp(f.mul(f.size("reads"), 4), f.gib(16), f.gib(128)), { fallback: "64GiB" })
    .cpuFormula(f.clamp(f.divCeil(f.size("reads"), f.gib(2)), 4, 32), { fallback: 16 })
    .software(sw).run()
```

## How it works

- **DSL + evaluator** (`exec/formula.lib.tengo`): pure AST constructors — `add/sub/mul/div/divCeil`, `max/min/clamp`, conditionals, unit constants, `size`/`lineCount` — and an integer-only evaluator. A formula must evaluate to a positive integer; a bool, negative, or zero result is rejected at run time, naming the offending dimension.
- **Metrics**: `f.size(tag?)` sums input blob sizes via `getBlobSize` (no pre-exec). `f.lineCount(tag?)` counts newlines via a per-file `line-counter` pre-exec that streams `.gz`/`.bz2`/`.zst`.
- **Evaluated in the pure template** (`exec/exec.tpl.tengo`): the formula AST rides a **meta** input — a value resource, readable as final like `quota` — and the line-count map rides a **regular** input, so the pure template's automatic readiness gate deep-waits it. No `awaitState`. The ephemeral impl just consumes the computed quota.
- **CID-transparent**: the formula AST is a meta input, so it is excluded from the exec CID — adding or changing a formula leaves dedup unchanged. The line-count map is a regular (in-CID) input, but it is file-derived, so identical inputs still dedup. Transparency needs backend `serviceFields`; without it the formula AST is attached as a regular input and enters the CID — evaluation stays correct, only dedup is lost.
- **Fallback**: on a backend without `getBlobSize`, each dimension falls back to its `{ fallback }` or a static `.mem()`/`.cpu()` — validated like a direct request — and a formula with neither raises a clear error.
- Also: `pt.wf.frame()` gains a `delimiter` option for csv/tsv (spec R19), forwarding to PTabler's `ReadCsvStep`.

## Key design decisions

- **`lineCount` runs a `line-counter` binary, not PTabler.** `:pt` imports `:exec`, so reaching a pt-based metric from `:exec` would close a compile-time import cycle, and DI can't break it (this Tengo fork re-instantiates modules per import). A small streaming binary sidesteps both.
- **Formulas evaluate in the pure template, CID-transparently.** The formula AST is a value resource, so a meta input carries it and it reads as final in the pure body — the same path `quota` uses. Only the computed line-count map needs a deep-waited regular input. `getBlobSize` is the only gate; `serviceFields` affects only CID-transparency, with a correct regular-input fallback when it is absent.

## Design history

This branch reached a tested, release-ready state, then continued as the design evolved:

- **`fcb94f0c`** — the prior PR head. **Option 1**: the formula evaluated in the *ephemeral* impl, CID-transparent via meta inputs. Tested and release-ready.
- **`601f4e05`** (+ `0bebb92`) — **Option 2**: moved evaluation into the *pure* template, with the formula AST and metrics on *regular* inputs, so the formula entered the exec CID (transparency dropped).
- **`f64a662b` → HEAD** — the **hybrid** (current design): evaluate in the pure template *and* keep the formula CID-transparent. The formula AST returns to a meta input; only the file-derived line-count map stays a regular input.

## Testing

- **Unit**: 191 tengo tests — DSL, evaluator, `resolveAst`, and the fallback resolver.
- **E2e** (17 cases, live backend): the formula → quota → `{system.ram.bytes}` chain is byte-exact (e.g. `1 GiB + 327` → `1073742151`), covering the `.gz`/`.bz2`/`.zst` lineCount paths, multi-file aggregation, the rejection guards, and CID behavior — `formula-cid-transparent` dedups two execs that differ only by formula to one allocation, while measuring a different file differentiates.

In addition to the automated tests, the three dependent blocks below were tested locally (with a smaller formula, so the allocation visibly changed across different input files on a local Mac).

## Status

Implementation, tests, and review fixes are complete. Merge follows:

- **Releasing `@platforma-sdk/workflow-tengo`** — this PR carries the `minor` changeset; the dependent blocks adopt the API once it ships.
- **Three dependent block PRs** (the reference-block migrations): platforma-open/mixcr-clonotyping#195, platforma-open/mixcr-amplicon-alignment#86, platforma-open/miltenyi-tcr-bcr-clonotyping#30.

`software-small-binaries@2.1.1` (the `line-counter` entrypoint) is released; the catalog pin moves `^2.0.3 → ^2.1.1`. The Phase 3 spec lives in milaboratory/text#151. PTabler-based blocks (e.g. import-vdj-data) come in Phase 4, which adds the formula API to `pt.workflow()`.

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds a resource-formula DSL to `exec.builder()` so block workflows can compute RAM and CPU requests from input-file metrics (byte size, newline count) at run time, replacing hard-coded `.mem()` / `.cpu()` values. The formula AST rides the META input rail (CID-transparent), while a computed per-file line-count map rides the REGULAR rail and is deep-waited to final in the pure exec template.

- **`exec/formula.lib.tengo`** — new pure DSL: AST constructors (`add/sub/mul/div/divCeil`, `max/min/clamp`, `gt/gte/lt/lte/and/or/not`, `if`, `size`, `lineCount`), `resolveAst`, `referencedMetricFiles`, `evaluate`/`evaluateResource`, and `applyFormulaFallback` for no-getBlobSize backends.
- **`exec/index.lib.tengo`** — adds `.memFormula()` / `.cpuFormula()` to `builder()`; `.addFile()` gains `{ tag }` option; `_lineCountResource` runs a streaming `line-counter` binary (no PTabler, no import cycle) and feeds results into an aggregate map resource; the GPU-resource guard is widened to accept either a static request or a formula.
- **`pt/index.lib.tengo`** — adds `delimiter` option to `pt.wf.frame()` for csv/tsv reads; 191 unit tests and 17 E2E cases (byte-exact quota verification, compressed-codec coverage, CID-transparency proofs, rejection-guard negative paths) accompany the changes.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge. The formula evaluation path is well-isolated (pure AST with no side effects), the hasBlobSize gate ensures graceful fallback on older backends, and the E2E suite verifies byte-exact quota allocation end-to-end.

The core formula DSL, evaluator, and exec-template wiring are all covered by 191 unit tests and 17 live-backend E2E cases, including negative-path rejection guards. Previously flagged issues (static .mem()/.cpu() as implicit fallback, the GPU-guard false-positive on hasBlobSize backends, and the lineCount map null-guard) are all addressed in this revision. The only open items are JSDoc omissions for the new { tag } option, which do not affect runtime behaviour.

No files require special attention. The exec.tpl / index.lib / formula.lib trio are all consistent, the constants are shared across attach/read/validate sites, and the line-counter pre-exec is correctly gated on the hasBlobSize feature flag.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| sdk/workflow-tengo/src/exec/formula.lib.tengo | New pure DSL + evaluator: AST constructors, resolveAst, referencedMetricFiles, evaluate/evaluateResource with positive-int guard, applyFormulaFallback. Well-tested (191 unit cases); the documented divCeil assumption (a ≥ 0) is noted but not enforced at the formula level — caller must use f.max/f.clamp, which is documented. |
| sdk/workflow-tengo/src/exec/index.lib.tengo | Adds memFormula/cpuFormula to builder, tag support on addFile, _lineCountResource/_aggregateLineCountMap helpers; GPU guard widened to accept formulaAst; applyFormulaFallback correctly treats existing static .mem()/.cpu() as implicit fallback on no-getBlobSize backends; JSDoc for addFile/addFiles omits the new { tag } option. |
| sdk/workflow-tengo/src/exec/exec.tpl.tengo | Formula evaluation relocated to the pure template, using getBlobSize for size metric and deep-waited line-count map for lineCount metric; quota cloned and overridden per-dimension; schema extended with optional formula/metric inputs via computed keys; CID-transparent via the META rail. |
| sdk/workflow-tengo/src/exec/exec-impl.tpl.tengo | Minimal change: validateInputs refactored to separate inputSchema variable (enables computed-key extension in exec.tpl); comment clarifies that quota is already fully computed before reaching here. |
| sdk/workflow-tengo/src/exec/constants.lib.tengo | Adds three string constants (META_INPUT_MEM_FORMULA, META_INPUT_CPU_FORMULA, REGULAR_INPUT_LINE_COUNT) as a single source of truth for the cross-file input-name contract. |
| sdk/workflow-tengo/src/pt/index.lib.tengo | Adds delimiter option to pt.wf.frame() for csv/tsv reads; validated and forwarded to ReadCsvStep; independent of the resource-formula path. |
| tests/workflow-tengo/src/exec/formula.test.ts | 17 E2E test cases covering: size/lineCount/multi metrics, compressed codecs (.gz/.zst/.bz2), CID-transparency proofs, rejection-guard negative paths, and zero-newline edge case; quota verified byte-exact via {system.ram.bytes},{system.cpu} echoed from the runner. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant B as exec.builder().run()
    participant F as exec.formula
    participant LC as _lineCountResource
    participant ET as exec.tpl (pure)
    participant EI as exec-impl.tpl (ephemeral)

    B->>F: resolveAst(memFormulaAst, tagMap, allFiles)
    F-->>B: "resolvedMem (metric -> metricSum)"
    B->>F: referencedMetricFiles(resolvedMem)
    F-->>B: "{ lineCount: [file.fastq] }"
    B->>LC: _lineCountResource(file.fastq, fileRef)
    LC-->>B: contentRef (newline count as text)
    B->>B: "_aggregateLineCountMap({ file.fastq: contentRef })"
    Note over B: resolvedMem -> META input __memFormula__
    Note over B: lineCountMap -> REGULAR input __metric_lineCount__
    B->>ET: "render(metaInputs={quota,__memFormula__}, pureInputs={filesToAdd,__metric_lineCount__})"
    Note over ET: Pure body runs once all inputs are final
    ET->>ET: ll.getBlobSize(fileRefs[fn].id)
    ET->>ET: lineCountMap[fn].getData()
    ET->>F: evaluateResource(resolvedMem, resolver, ram)
    F-->>ET: computed RAM bytes (positive int)
    ET->>ET: "quota.ram = computed + b"
    ET->>EI: "render(metaInputs={quota(computed)}, pureInputs={mainPlan,...})"
    EI->>EI: wdBuilder.cpu(quota.cpu).mem(quota.ram)
    EI-->>ET: exec outputs
    ET-->>B: exec result
```
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
sdk/workflow-tengo/src/exec/index.lib.tengo:898-901
The `{ tag }` option added in this PR is not mentioned in the JSDoc for `addFile` or `addFiles`. A block author reading the docs won't know the option exists until they find it in the source.

```suggestion
		 * @param opts: { writable?: bool, tag?: string, mnz?: ... } - optional flags.
		 *        `writable: true` lands the file as 0o600 in the workdir.
		 *        `tag: "name"` groups the file with others under the same tag for
		 *        f.size("name") / f.lineCount("name") metric aggregation in resource formulas.
		 */
		addFile: func(fileName, file, ...opts) {
```

### Issue 2 of 2
sdk/workflow-tengo/src/exec/index.lib.tengo:923-926
`addFiles` forwards opts to `addFile`, so it also silently accepts `{ tag }` today, but its JSDoc still describes only `writable` and `mnz`. Block authors applying a single tag to a batch of files via `addFiles` will be surprised it works — or won't know it's possible.

```suggestion
		 * @param filesMap: map[string]reference - a map of file names to files' references.
		 * @param opts: { writable?: bool, tag?: string, mnz?: ... } - applied to every entry.
		 */
		addFiles: func(filesMap, ...opts) {
```

`````

</details>

<sub>Reviews (7): Last reviewed commit: ["refactor(workflow-tengo): tidy resource-..."](https://github.com/milaboratory/platforma/commit/d9faa3298721d79917af85fe75ee0148efa3569d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35431810)</sub>

### CODE REVIEW

**Comment by PaulNewling** [comment]
@greptileai

**Comment by PaulNewling** [comment]
@greptileai

**Comment by PaulNewling** [comment]
@greptileai

**Comment by PaulNewling** [comment]
@greptileai

File: assets/formula_no_newline.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+one line, no trailing newline
```
Comment (PaulNewling) [inline assets/formula_no_newline.txt:1]: These files in the `assets/` directory are used for testing to ensure the `f.linecount()` works as expected with different formats and edge cases.
There is already a couple other files like `maybe_the_number_of_lines_is_the_answer.txt` that are in `assets/` so this is just making sure we have full coverage

File: sdk/workflow-tengo/src/exec/formula.test.tengo [hunk 2]
```diff
@@ -0,0 +1,348 @@
+test := import(":test")
+f := import(":exec.formula")
```
Comment (PaulNewling) [inline sdk/workflow-tengo/src/exec/formula.test.tengo:2]: Since `formula.lib` has basically no dependencies, we can unit test just about every part of it to ensure its correctness.

**Comment by PaulNewling** [comment]
@greptileai

**Comment by PaulNewling** [comment]
@greptileai

File: sdk/workflow-tengo/src/exec/formula.lib.tengo [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+// bounds:     { op: "max"|"min", a, b } | { op: "clamp", value, lo, hi }
+// compare:    { op: "gt"|"gte"|"lt"|"lte", a, b }
+// logic:      { op: "and"|"or", a, b } | { op: "not", a }
+// cond:       { op: "if", cond, then, else }
+// metric:     { op: "metric", metric: "size"|"lineCount", "tag,?": <string>, "default,?": <int> }
+//             (the builder later resolves "metric" -> "metricSum" with concrete files)
+// Note: if.cond and not.a may be const-wrapped if given a bare int; the
+// evaluator treats any nonzero const as truthy.
+
+GIB := 1073741824
```
Comment (vadimpiven) [inline sdk/workflow-tengo/src/exec/formula.lib.tengo]: we do have these already in units, why not to re-export?

File: sdk/workflow-tengo/src/exec/formula.lib.tengo [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+MIB := 1048576
+KIB := 1024
+GB  := 1000000000
+MB  := 1000000
+KB  := 1000
+
+// Ops _wrap accepts as valid expression nodes. "metricSum" is builder-internal
+// (resolveAst rewrites "metric" -> "metricSum" with concrete files); the DSL
+// never emits it directly.
+_KNOWN_OPS := {
```
Comment (vadimpiven) [inline sdk/workflow-tengo/src/exec/formula.lib.tengo]: why not sets.fromSlice ?

**Review by vadimpiven — APPROVED** [review]
